### ISSUE TITLE
Add EXT2 FS

### ISSUE THREAD

**Issue by liqinggd** [asterinas/asterinas#346]
This issue is used to track how to implement an Ext2 file system and add it to Asterinas.

# Status
Here we need to summarize the current status of the Ext2 file system implementation and the features that need to be implemented next, in order to guide our future work.
- [x] (**Functionality**)Define the relevant data structures and methods that comply with the Ext2 file system specification, enabling correct reading and parsing of disk images in the Ext2 format.
- [x] (**Functionality**)Implement the basic functionalities of the fs, including management and querying of inodes and blocks, file read/write, directory operations, and symbolic links.
- [x] (**Functionality**)Remove the unsafe code.
- [x] (**Functionality**)Implement regular files, directories, and symlinks with page cache.
- [x] (**Functionality**)Support indirect blocks, double indirect blocks, and triple indirect blocks in order to support large files.
- [x] (**Functionality**)Implement a simple BIO layer based on a queue mechanism to facilitate the connection between the file system and the block device driver(i.e., `virtio-blk`).
- [x] (**Functionality**)Figure out how to configure `virtio-blk` to directly use an ext2 formatted file as the backend storage.
- [x] (**Reliability**)Implement a comprehensive consideration and resolution of concurrent access issues. (WIP)
- [x] (**Performance**)Add a cache mechasim to cache the metadata of inode, reducing the number of small I/O operations.
- [x] (**Performance**)Pass some fs benchmarks(e.g., `fio`, `filebench`)
- [x] (**Functionality**)Enable Gvisor syscall test.

# Core new components
- The Ext2 FS
The FS itself must implement the core functionality of Ext2. It can open an ext2-formated block device and manage the superblock, blocks, and inodes from it.

- The Block I/O layer
The block I/O layer should provide an abstraction of the block device and the basic API for submitting I/O requests to the block device.

# Ext2

## The Design principle of Ext2
 - No unsafe code
 - Use page cache
 - Use the new BIO layer

### No unsafe code
We should provide some components in the frame to accomplish all the unsafe work and ensure memory safety.

For example,
- Read or write the pod metadata (superblock, block group descriptor, inode) from or to the block device.
- Read or write a non-pod struct dir entry from or to the block device.
- Submit Bio request to block device.

### Use page cache

It depends on #419
 
Currently, we implement page cache in the `vnode` layer, so the page cache is transparent to `dyn Inode` trait. The page cache only handles the `read` and `write` methods, so it is a good design to decrease the complexity of each inode's implementation.

Ext2 uses page cache in `File`, `Dir`, and `Symlink`. Ext2 inode may read, write, or resize the page cache in most implementations of the methods of `dyn Inode` trait, so it seems that Ext2 inode should manage the page cache for itself.

For example, the `create` method for Ext2 inode may search the page cache and write a dir entry. If the page cache is too small, it may expand the page cache to append the dir entry. If we put the page cache above the implementation of Ext2 inode, the `create` method has to add the page cache as another parameter. Such a case may happen to many methods of `dyn Inode` trait.
```rust
fn create(&self, name, type_, mode, pagecache);
fn lookup(&self, name, page_cache);
fn unlink(&self, name, page_cache);

... ...

```

### Use the new BIO layer
The Ext2 should use the new APIs of the BIO layer.

#### Standalone crate

This feature depends on issue #418

Pros.
It is easy to implement many test cases for the crate itself and the crate can run test cases in user mode.

Cons.
we need to add an adaptor layer to connect Ext2 with `aster-std`, and it makes the code sophisticated. For example, we can tailor some modules from the `aster-std`,  both `aster-std` and `ext2` crates depend on the new crate. It is difficult to tailor the `aster-std`.

#### Integrated into `aster-std`

Pros.
It is easy to implement Ext2 for Asterinas, because `aster-std` provides many useful VFS components.

Cons.
It is hard to test the Ext2 fs in use space. The `aster-std` becomes bigger.

### 2. How to use the block device

Please see #420 

# The Block I/O layer

Design API for submitting bio requests to block devices.
